### PR TITLE
Data Fusion API

### PR DESCRIPTION
**Pull request by joelbecker**
# Data Fusion for `recordlinkage`

This pull request introduces a new section of the `recordlinkage` API for merging matched records into a new data frame. This PR implements the following:

* A suite of conflict resolution functions for handling conflicts based on observed values, metadata, and user specifications (see: `recordlinkage/algorithms/conflict_resolution.py`)
* The `FuseCore` abstract class which implements features common to all data fusion use cases, including a general-purpose framework for handling data conflicts and implementations of several common conflict resolution strategies (see: `recordlinkage/fusion.py`).
* The `FuseLinks` class, which implements data fusion for two data frames (see: `recordlinkage/fusion.py`).

# Example Usage

Here is an example demonstrating how to fuse two data frames (using `load_febrl4` as in [linking two datsets](http://recordlinkage.readthedocs.io/en/latest/notebooks/link_two_dataframes.html)). To run the script, download the full code [here](https://gist.github.com/joelbecker/fa6f29ad19790cfaa4c9adf14455e2b1).

```python
# Fusion step
fuse = rl.FuseLinks()

# Prefer values in dataframe a
fuse.trust_your_friends('given_name', 'given_name', trusted='a', name='given_name')

# Choose values from the row that was updated most recently
fuse.keep_up_to_date('surname', 'surname', 'dates_updated', 'dates_updated', name='surname')

# Take the average of salary values
fuse.meet_in_the_middle('salary', 'salary', metric='mean', name='salary')

# Choose randomly between street numbers
fuse.roll_the_dice('street_number', 'street_number', name='street_number')

# Keep all social security id values for future processing.
fuse.pass_it_on('soc_sec_id', 'soc_sec_id', name='soc_sec_id')

# Handle data conflicts between multiple columns in each data frame
fuse.meet_in_the_middle(['min', 'max'], ['min', 'max'], metric='stdev', name='spread')

# Create custom conflict handling strategies with the resolve method
fuse.resolve(
    cr.choose_longest,
    ['address_1', 'address_2'],
    ['address_1', 'address_2'],
    tie_break=cr.choose_random,
    name='longest_address'
)

# Execute the scheduled conflict resolution jobs for the given
#   candidate links, data, and classifications.
fused = fuse.fuse(pairs, dfA, dfB, matches)
```

# Notes

## Data Fusion for Deduplication

As discussed in the relevant issue page, data fusion for deduplication has a different set of challenges than data fusion for dataframe linking. The `FuseCore` class was designed to be general enough to implement both `FuseLinks` and `FuseDuplicates`, but at this time only `FuseLinks` has been implemented. `FuseDuplicates` will require (a) clustering algorithms to determine groups of records to be fused into a new row, and (b) implementing `FuseDuplicates._make_resolution_series` to process data from grouped records into the format required by `FuseCore`.

## Refining Link Mapping

We also discussed implementing a feature which would "refine" the set of candidate links to enforce 1-to-n, m-to-1, or 1-to-1 link mappings. I chose not to implement this feature as a part of `FuseLinks`. This was partly to manage the scope of this pull request, and partly because I think that mapping refinement is an aspect of the _classification_ step, not the data fusion step.

## A Non-Fusion Change

I made a small change to fix a mistake in the documentation for the smith-waterman comparison function.

## Code Review & Questions

This is a fairly large pull request, and I can't explain all the specifics of the data fusion classes in this description. Please ask as many questions as you want!

## Acknowledgements

@jillianderson8 was instrumental in designing the data fusion API implemented in this pull request.

### CODE REVIEW

File: recordlinkage/fusion.py [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+            The name of the resulting resolved column.
+
+        Returns
+        -------
+        None
+            Queues a conflict resolution job for later computation, to be completed when ``fuse`` method is called.
+        """
+        self.resolve(no_gossip, values_a, values_b, name=name, remove_na_vals=False, description='no_gossiping')
+
+    def roll_the_dice(self, values_a, values_b, name=None, remove_na_vals=True):
```
Comment (J535D165) [inline recordlinkage/fusion.py]: The method names are funny to me. I should have named this 'random'. Is it common to name algorithms like these in such a way is academic literature about data fusion?

**Comment by J535D165** [comment]
Thanks a lot for this PR @joelbecker and @jillianderson8. This must have taken a lot of time. The design of the API is very nice. The algorithms look very useful. 

I do have a few remarks and I am still investigating some things. But it is a very large amount of code so I need some time. I am definitely going to merge this PR soon. If I can do something to help, let me know. 

I need to dive into the code a bit more to understand why `FuseLinks` and `FuseDuplicates` are separated. I prefer to have the same class name for both linking and deduplication (to keep the api simple). But I am not sure if this is inevitable for data fusion. This is closely related to the arguments of `FuseLinks.fuse()`. See also `Compare.compare()`

I see that you are still working on this PR. Can you let me know when it is ready to merge for you?

Again many thanks, Jonathan

File: recordlinkage/fusion.py [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+        )
+
+        data = data.apply(job['fun'], args=job['params'])
+
+        if job['name'] is not None:
+            data = data.rename(job['name'])
+
+        return data
+
+    def fuse(self, index, df_a, df_b, predictions=None, njobs=1, sep='_'):
```
[the remark this answers is not in the document]
Comment (joelbecker) [reply recordlinkage/fusion.py]: Excellent point. I'll tweak this.

File: recordlinkage/fusion.py (hunk 1 above)
Comment (joelbecker) [reply recordlinkage/fusion.py]: You're right on both counts — the method names are from the academic data fusion literature, but they're also a little funny. Jillian and I have been using the academic names, but I'm happy to change them if that's your preference. As I see it, the pros and cons of the current naming scheme are:

Pros: 
* Consistent with the academic literature.
* Distinguishes between conflict resolution strategies (e.g. `roll_the_dice`) and the conflict resolution functions that implement those strategies (e.g. `choose_random`).

Cons:
* The names are odd and require the user to learn what they mean, which increases the barrier to entry.

If you'd like them changed, let me know.

File: versioneer.py [hunk 3]
```diff
@@ -65,7 +65,7 @@
 The version identifier is used for multiple purposes:
 
 * to allow the module to self-identify its version: `myproject.__version__`
-* to choose a name and prefix for a 'setup.py sdist' tarball
+* to choose_trusted a name and prefix for a 'setup.py sdist' tarball
```
[the remark this answers is not in the document]
Comment (joelbecker) [reply versioneer.py]: My bad. Not sure what happened, but probably a refactoring mistake.

**Comment by joelbecker** [comment]
Thanks for your comments, Jonathan! You're right, I'm still working on making the PR backwards-compatible to Python 2.7. I'll let you know as soon as it is ready to merge.

There were a few reasons for separating `FuseLinks` and `FuseDuplicates`. The main reason was that while `FuseLinks` can fuse records as-is based on the candidate links and classifications from previous steps, `FuseDuplicates` requires records to be grouped into clusters first. This is significant, because the pre-processing to format data for the conflict resolution step will be more involved for clustered records. I wanted a single conflict handling framework that would handle both, but wanted to totally separate the data-formatting step to avoid adding unnecessary performance overhead in `FuseLinks` to accommodate  for duplicate clustering. So, I designed the class hierarchy with this in mind:

* `FuseCore` assumes that it will be given data in a consistent format. It implements data handling strategies, and coordinates queueing and executing resolution jobs.
* The subclasses handle data formatting, applying predictions to the set of candidate links (`MultiIndex`) and any features that apply to one case and not the other.

`FuseDuplicates` hasn't been implemented yet, but these are the differences I would anticipate between the two subclasses:

`FuseLinks`
* `trust_your_friends`/`choose_trusted` is implemented in `FuseLinks` since it only makes sense when there are two data sources.
* `_make_resolution_series` formats data by zipping together series made with the index and `iloc`.
* Initialized as `FuseLinks(njobs = 1, sep = '_')`

`FuseDuplicates`
* Should implement at least one, but ideally several clustering algorithms.
* May include conflict handling strategies unique to deduplication.
* `_make_resolution_series` will be significantly complicated by resolving conflict among a cluster of records instead of a pair of records.
* Would be initialized as `FuseDuplicates(cluster_method='method_name', njobs=1, sep='_')`.

I feel that there are a number of benefits of this design:
* This design avoids the confusion of identifying which features are relevant to the users' linking or deduplicating problem. So, a user linking two data frames doesn't need to worry about `cluster_method` or mistakenly think that data clustering is occurring when linking two data frames.
* This feels like a nice object oriented design. I prefer overriding the `_make_resolution_series` method, rather than including a pair of situation-dependent formatting options.

Having said this, these could easily be merged into a single class:
* Alternate versions of `_make_resolution_series` would need to be implemented. Depending on the "mode" being used, one of the two methods would be used for data processing.
* Warnings/errors if the user tries to use an inappropriate method (e.g. `trust_your_friends` in deduplication) or unsupported option (e.g. specifying a clustering algorithm in linking).

If you feel that avoiding separating linking/deduplication into different classes is a big problem, we can talk about integrating them (when data fusion for deduplication is implemented). However, I think that separating them is a much cleaner approach. I also think that data fusion has greater differences between the approaches needed for deduplication and linking compared to previous steps.

Let me know your thoughts!

(I think that a key part of this are the mechanics of how `_make_resolution_series` processes data, and the format expected by `FuseCore`. I'll write a follow-up post, since I didn't want to keep this one from getting too long.)

**Comment by joelbecker** [comment]
This post is a run-down of how the data fusion classes format data for conflict resolution. I've done my best to explain these things in doc strings & comments, but it's pretty central to the way these classes work, so I wanted to write it up explicitly.

# Format

`FuseCore` expects data in a specific format. This format was chosen with the following constraints:

* As efficient as possible.
* Highly flexible, appropriate for both data linking and deduplication. Specifically, the format should be able to handle any number of data values, since clusters may be arbitrarily large.

So, instead of passing a set number of values, `FuseCore` expects a set of values, with an optional piece of metadata paired with each value. These are passes as a a tuple of values and a tuple of metadata values. In the case of a value-only resolution strategy (such as `roll_the_dice` or `cry_with_the_wolves`) the data will be formatted in this way:

```python
tuple(tuple(v1, ..., vn), )
```

When a strategy requires metadata (e.g. `keep_up_to_date`), a second tuple is included:

```python
tuple(tuple(v1, ..., vn), tuple(m1, ..., mn))
```

Where `vi`, `mi` are a value/metadata pair.

# `_make_resolution_series`

`_make_resolution_series` creates a series of these nested tuples, and `FuseCore` applies a conflict resolution function to the series as specified by the user. `_make_resolution_series` will process data differently for data linking and data duplication.

## `FuseLinks` Example

In `FuseLinks`, the appropriate values are included from the appropriate pair of rows.

**DataFrame 1**
<table>
<thead>
<tr><th style="text-align: right;">  </th><th>a    </th><th>b    </th></tr>
</thead>
<tbody>
<tr><td style="text-align: right;"> 0</td><td>d1.a0</td><td>d1.b0</td></tr>
<tr><td style="text-align: right;"> 1</td><td>d1.a1</td><td>d1.b1</td></tr>
<tr><td style="text-align: right;"> 2</td><td>d1.a2</td><td>d1.b2</td></tr>
</tbody>
</table>

**DataFrame 2**
<table>
<thead>
<tr><th style="text-align: right;">  </th><th>a    </th><th>b    </th></tr>
</thead>
<tbody>
<tr><td style="text-align: right;"> 0</td><td>d2.a0</td><td>d2.b0</td></tr>
<tr><td style="text-align: right;"> 1</td><td>d2.a1</td><td>d2.b1</td></tr>
<tr><td style="text-align: right;"> 2</td><td>d2.a2</td><td>d2.b2</td></tr>
</tbody>
</table>

If we call `_make_resolution_series('a', 'a', meta_a='b', meta_b='b')`, the corresponding data for a candidate link `(1, 2)`
would be:

```
tuple(tuple('d1.a1', 'd2.a0'), tuple('d1.b1', 'd2.b0'))
```

Note that the user can specify an arbitrary number of columns from each data frame to resolve. Hypothetically, if we
added a `'c'` column and instead called `_make_resolution_series(['a', 'c'], ['a', 'c'], meta_a='b', meta_b='b')` we
could get:

```
tuple(tuple('d1.a1', 'd1.c1', 'd2.a0', 'd2.c0'), tuple('d1.b1', 'd1.b1', 'd2.b0', 'd2.b0'))
```

## `FuseDuplicates` Example

`FuseDuplicates` is somewhat more complicated. In the `FuseLinks` example, data may be included from any number of columns,
but only ever a single pair of data frame rows. However, because we're fusing clusters of records, we might have an arbitrary
number of columns and rows which should be included. So, if we have the dataframe:

**DataFrame 3**

<table>
<thead>
<tr><th style="text-align: right;">  </th><th>a    </th><th>b    </th><th>c    </th></tr>
</thead>
<tbody>
<tr><td style="text-align: right;"> 0</td><td>d3.a0</td><td>d3.b0</td><td>d3.c0</td></tr>
<tr><td style="text-align: right;"> 1</td><td>d3.a1</td><td>d3.b1</td><td>d3.c1</td></tr>
<tr><td style="text-align: right;"> 2</td><td>d3.a2</td><td>d3.b2</td><td>d3.c2</td></tr>
<tr><td style="text-align: right;"> 3</td><td>d3.a3</td><td>d3.b3</td><td>d3.c3</td></tr>
</tbody>
</table>

And hypothetically call (not yet implemented) `FuseDuplicates._make_resolution_series('a', meta='b')`,
the data corresponding to the cluster `{0,2,3}` would be:

```python
tuple(tuple('d3.a0', 'd3.a2', 'd3.a3'), tuple('d3.b0', 'd3.b2', 'd3.b3'))
```

And if we called `FuseDuplicates._make_resolution_series(['a', 'c'], meta='b')` we would get:

```python
tuple(
    tuple('d3.a0', 'd3.a2', 'd3.a3', 'd3.c0', 'd3.c2', 'd3.c3'),
    tuple('d3.b0', 'd3.b2', 'd3.b3', 'd3.b0', 'd3.b2', 'd3.b3')
)
```

File: recordlinkage/fusion.py (hunk 1 above)
Comment (J535D165) [reply recordlinkage/fusion.py]: Clear! I prefer the academic literature on data fusion. I do have to read more about data fusion (I am using record linkage techniques for population estimation most of the time).

File: versioneer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (J535D165) [reply versioneer.py]: No problem

**Comment by joelbecker** [comment]
I also fixed the typo in `versioneer.py`.

**Comment by J535D165** [comment]
Commit 2f7ed457266dbac28b2d960d196759a4e242a52f adds a method for clustering data (ConnectedComponents). Also some other methods relevant for data fusion like One-to-One Linking. I will merge this PR and will take a look at the deduplication part.

**Comment by joelbecker** [comment]
Excellent! Thanks for merging, and I'll look forward to digging into these new features!

**Pull request merged**